src/Simplex/Chat/Library/Commands.hs

```diff
@@ -3169,46 +3228,44 @@ processChatCommand' vr = \case
                 quoteData ChatItem {content = CISndMsgContent qmc} = pure (qmc, CIQDirectSnd, True)
                 quoteData ChatItem {content = CIRcvMsgContent qmc} = pure (qmc, CIQDirectRcv, False)
                 quoteData _ = throwError SEInvalidQuote
-    sendGroupContentMessages :: User -> GroupInfo -> Maybe GroupMemberId -> Bool -> Maybe Int -> NonEmpty ComposedMessageReq -> CM ChatResponse
-    sendGroupContentMessages user gInfo@GroupInfo {membership} directMemId_ live itemTTL cmrs = do
+    sendGroupContentMessages :: User -> GroupInfo -> GroupChatScope -> Bool -> Maybe Int -> NonEmpty ComposedMessageReq -> CM ChatResponse
+    sendGroupContentMessages user gInfo gcs live itemTTL cmrs = do
       assertMultiSendable live cmrs
-      (ms, numFileInvs, notInHistory_) <- case directMemId_ of
-        Nothing -> do
-          ms <- withFastStore' $ \db -> getGroupMembers db vr user gInfo
-          pure (ms, length $ filter memberCurrent ms, Nothing)
-        Just dmId -> do
-          when (dmId == groupMemberId' membership) $ throwChatError $ CECommandError "cannot send to self"
-          dm <- withFastStore $ \db -> getGroupMemberById db vr user dmId
-          unless (memberStatus dm == GSMemPendingApproval) $ throwChatError $ CECommandError "cannot send directly to member not pending approval"
-          pure ([dm], 1, Just NotInHistory)
-      sendGroupContentMessages_ user gInfo notInHistory_ ms numFileInvs live itemTTL cmrs
-    sendGroupContentMessages_ :: User -> GroupInfo -> Maybe NotInHistory -> [GroupMember] -> Int -> Bool -> Maybe Int -> NonEmpty ComposedMessageReq -> CM ChatResponse
-    sendGroupContentMessages_ user gInfo@GroupInfo {groupId, membership} notInHistory_ ms numFileInvs live itemTTL cmrs = do
-      -- TODO [knocking] pass GroupSndScope?
-      let allowedRole = case ms of
-            [m] | memberCategory m == GCHostMember && memberStatus membership == GSMemPendingApproval -> Nothing
-            _ -> Just GRAuthor
+      (gcsi, msgScope, recipients, numFileInvs) <- getGCSDependencies vr user gInfo gcs modsCompatVersion
+      sendGroupContentMessages_ user gInfo gcs gcsi msgScope recipients numFileInvs live itemTTL cmrs
+        where
+          hasReport = any (\(ComposedMessage {msgContent}, _, _, _) -> isReport msgContent) cmrs
+          modsCompatVersion = if hasReport then contentReportsVersion else groupKnockingVersion
+    sendGroupContentMessages_ :: User -> GroupInfo -> GroupChatScope -> GroupChatScopeInfo -> MsgScope -> [GroupMember] -> Int -> Bool -> Maybe Int -> NonEmpty ComposedMessageReq -> CM ChatResponse
+    sendGroupContentMessages_ user gInfo@GroupInfo {groupId, membership} gcs gcsi msgScope recipients numFileInvs live itemTTL cmrs = do
       forM_ allowedRole $ assertUserGroupRole gInfo
       assertGroupContentAllowed
       processComposedMessages
       where
+        allowedRole :: Maybe GroupMemberRole
```

Comment:
why is it moved to the bottom? return it back.

Reply:
also, I think we should allow support chats to observers. And not check role for host, as before

Reply:
because it became bigger and unnecessarily breaks the flow of function